### ISSUE TITLE
CRITICAL HARD services (because of host down) recover SOFTly.

### ISSUE THREAD

**Issue by ssaxnagios** [NagiosEnterprises/nagioscore#759]
CRITICAL HARD services (because of host down) recover as SOFT state (BUG).

Validated as bug in 4.4.5.

- Host goes DOWN in SOFT state
- Service checks goes HARD immediately because of host
- Host recovers (SOFT recovery)
- Service recovers (SOFT recovery, should be HARD)

That service not returning to HARD affects the availability reports.

**Comment by nicomatsu** [comment]
For sake of completeness, this issues arises from https://support.nagios.com/forum/viewtopic.php?f=16&t=58360

As I told @ssaxnagios if the proposed correction is implemented it poses a big risk of no notification in the current example:

- HOST enters SOFT DOWN (attempt 1 of 3, suppose max_check_attemps = 3)
- SERVICE enters HARD CRITICAL (attempt 1 of 5, suppose max_check_attemps = 5) with NO NOTIFICATION because host is down
- HOST check returns UP, it is a SOFT recovery
- SERVICE check continues to give CRITICAL because it was a real failure and no notfication will be sent

The underlying bug is that the service notification was suppressed by a host non-UP SOFT state.
On the other hand, if notification is not suppressed then the so good host down suppressing service notifications is gone.

This behaviour that is wanted is that service goes HARD state after host goes non-UP HARD state.
What I am describing is not my idea but the behaviour of Nagios Core 4.3.4 as shown in my tests in https://support.nagios.com/forum/viewtopic.php?f=16&t=58360

```
[1588633155] HOST ALERT: test;DOWN;SOFT;1;CRITICAL test
[1588633171] SERVICE ALERT: test;Manual test service;CRITICAL;SOFT;1;CRITICAL test
[1588633171] HOST ALERT: test;DOWN;SOFT;2;CRITICAL test
[1588633231] HOST ALERT: test;DOWN;HARD;3;CRITICAL test
[1588633231] HOST NOTIFICATION: nagiosadmin;test;DOWN;notify-host-by-email;CRITICAL test
[1588633291] SERVICE ALERT: test;Manual test service;CRITICAL;HARD;3;CRITICAL test
```

**Comment by sawolf** [comment]
Thanks for reporting this. I'll look into it the next time we do a bugfix release for Core.

**Comment by cdienger** [comment]
+1 the service not sending a notification after recovering into a soft state. See https://support.nagios.com/forum/viewtopic.php?f=16&t=60189. 

[1600855200] HOST DOWNTIME ALERT: hostX;STARTED;
[1600855200] HOST NOTIFICATION: contactX;hostX;DOWNTIMESTART

[1600855881] HOST ALERT: hostX;DOWN;SOFT;1;
[1600855970] HOST ALERT: hostX;DOWN;SOFT;2;
[1600856057] SERVICE ALERT: hostX;serviceX;CRITICAL;HARD;1
[1600855970] HOST ALERT: hostX;DOWN;SOFT;3;
...
[1600856681] HOST ALERT: hostX;DOWN;HARD;10;

[1600858799] HOST DOWNTIME ALERT: hostX;STOPPED;

[1600858799] HOST NOTIFICATION: contactX;hostX;DOWNTIMEEND
[1600858867] HOST NOTIFICATION: contactX;hostX;DOWN

[1600861318] HOST NOTIFICATION: contactX;hostX;UP
[1600861318] HOST ALERT: hostX;UP;HARD;

[1600861387] SERVICE NOTIFICATION: contactX;hostX;serviceX;

**Comment by workingninja** [comment]
This is also happening in 4.4.3 (at least I think this is the same issue). What is most confusing on my end is that the "critical" service notifications are sent with no "recovery" service notifications:

> [1602926691] HOST ALERT: host;DOWN;SOFT;1;(Host check timed out after 30.01 seconds)
> [1602926717] SERVICE ALERT: host;service;CRITICAL;HARD;1;connect to address 1.2.3.4 and port 12345: No route to host
> 
> [1602926754] HOST ALERT: host;DOWN;SOFT;2;CRITICAL - Host Unreachable (1.2.3.4)
> 
> [1602926814] HOST NOTIFICATION: nagiosadmin;host;DOWN;notify-host-by-email;CRITICAL - Host Unreachable (1.2.3.4)
> [1602926814] HOST ALERT: host;DOWN;HARD;3;CRITICAL - Host Unreachable (1.2.3.4)
> 
> [1602926998] HOST NOTIFICATION: nagiosadmin;host;UP;notify-host-by-email;PING OK - Packet loss = 0%, RTA = 0.59 ms
> [1602926998] HOST ALERT: host;UP;HARD;1;PING OK - Packet loss = 0%, RTA = 0.59 ms
> 
> [1602927016] SERVICE NOTIFICATION: nagiosadmin;host;service;CRITICAL;notify-service-by-email;service: not running
> 
> [1602927315] SERVICE ALERT: host;service;OK;SOFT;1;service: Running

There are no further entries in the log for the host service.

Please let me know if there is anything else I can do to help. Thanks!

**Comment by rousseaubenjamin** [comment]
Is there any work around to this ?

I have my services going through PagerDuty, and some stay stuck on Pagerduty without being auto resolve because of this bug. Thank you :)

**Comment by heavenlyren** [comment]
Has this been addressed in Nagios XI 5.8.1 ? ( I don't believe so because nagios core has not been updated )

**Comment by Ysincit** [comment]
I think, I found the main cause of this bug in line 1392 of base/checks.c (based on source of 4.4.6).
`1392:			svc->current_attempt = 1;`
even sets current_attempt to 1 if it was forced to be max_attempts previously in line 1354 (and HARD state in line 1352) because of host != HOST_UP :
`1352:                       svc->state_type = HARD_STATE;`
`1354:                       svc->current_attempt = svc->max_attempts;`
as consequence of the unexpected inconsistent combination of HARD state and current_attempt < max_attempts, some follow up logic works wrong.

Possible solution:
introduce a new flag variable.
`	int forced_hard          = FALSE;  `
`1354:                       svc->current_attempt = svc->max_attempts;  forced_hard = TRUE;`
`1392:			if (! forced_hard) svc->current_attempt = 1;`

An additional bug is, that 1354: svc->current_attempt = svc->max_attempts does not check if the service is going from a problem state to OK while the host is DOWN, and yes, this is possible, and will result in the inconsistent combination of HARD STATE_OK with svc->current_attempt = svc->max_attempts, so 1354 should be:
`1354:                       if (svc->current_state == STATE_OK) {  svc->current_attempt = 1; } else { svc->current_attempt = svc->max_attempts;  forced_hard = TRUE; }`

Best regards

**Comment by sawolf** [comment]
After fixing #837, this is the behavior I'm getting (forced check lines omitted):
```
[1667502498] HOST ALERT: localhost;DOWN;SOFT;1;PING DOWN
[1667502500] SERVICE ALERT: localhost;PING;CRITICAL;HARD;1;PING DOWN
[1667502510] HOST ALERT: localhost;DOWN;SOFT;2;PING DOWN
[1667502538] SERVICE ALERT: localhost;Current Users;CRITICAL;HARD;1;USERS situation is BAD
[1667502567] HOST FLAPPING ALERT: localhost;STARTED; Host appears to have started flapping (20.7% change > 20.0% threshold)
[1667502567] HOST ALERT: localhost;UP;SOFT;1;PING UP
[1667502588] SERVICE ALERT: localhost;Current Users;OK;HARD;1;USERS situation is OK
```
which looks correct to me. Is anyone else able to confirm that this issue is fixed in `master`?

**Comment by ssaxnagios** [comment]
I just tested master and it does look to be fixed. Thanks @sawolf!
